Remove fgrad_input from slow_conv2d

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64280

Differential Revision: [D30830887](https://our.internmc.facebook.com/intern/diff/D30830887)

peterbell10 committed:
Remove fgrad_input from slow_conv2d

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove fgrad_input from slow_conv2d"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove fgrad_input from slow_conv2d"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove fgrad_input from slow_conv2d"




[ghstack-poisoned]

ezyang commented:
ooh so delightful

cc @jbschlosser

ezyang commented:
Actually, @jbschlosser, I'm going to leave reviewing and landing this one up to you, since it is a logical step towards convolution rationalization

jbschlosser commented:
Very timely PR, @peterbell10 :) love it!

jbschlosser commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -369,6 +341,11 @@ static void slow_conv2d_backward_parameters_out_cpu_template(
   auto input = input_.contiguous();
   auto grad_output = grad_output_.contiguous();
 
+  if (grad_bias.defined()) {
+    Tensor grad_bias_mut = grad_bias;
+    at::sum_out(grad_bias_mut, grad_output, IntArrayRef{0, 2, 3});
+  }
+
```
This is a much nicer way to compute `grad_bias` and seems correct AFAICT.

Couple questions regarding this:
* Are there any tests verifying the `grad_bias` calculation? I briefly checked but nothing stood out
* Do you by chance have any comparative benchmarking results for the change?

peterbell10 replied:
>Are there any tests verifying the grad_bias calculation? I briefly checked but nothing stood out

The nn tests are a bit confusing, but I think the gradcheck is done here:
https://github.com/pytorch/pytorch/blob/59fcbd172b5dfdeb12e2f1b7a78c2ce95c1eb680/torch/testing/_internal/common_nn.py#L5807

That test includes inputs and parameters, so `bias` should be covered.


>Do you by chance have any comparative benchmarking results for the change?

I've done some simple benchmarking which showed positive results. LMK if you want more. 

For CUDA, I've run a convolution in a loop under nvprof with these parameters:
```
input = torch.rand(1, 50, 260, 260, device=device, requires_grad=False)
weight = torch.rand(10, 50, 3, 3, device=device, requires_grad=False)
bias = torch.rand(10, device=device, requires_grad=True)
```
The result is it calls one kernel instead of two (cublas gemm here becomes `dot_kernel` + `reduce_1Block_kernel`). The kernel execution time is also noticeably faster:

|          | Master (us) | This PR (us) |
|----------------:|:-----------:|:--------------------:|
|  Forward |     23.8    |         14.6         |
| Backward |     22.4    |         18.1         |

On CPU without nvprof, the results are dwarfed by the main convolution and the same shapes showed no measurable difference.  So I also tried a smaller input size with H=120, C_in = 2 and that showed some minor improvement in the forward case (presumably from vectorization).

|                  | Master (us) | This PR (us) |
|-----------------:|:-----------:|:--------------------:|
|    Forward (Big) |    15,100   |        15,100        |
|   Backward (Big) |    13,100   |        13,100        |
|  Forward (Small) |     526     |          499         |
| Backward (Small) |    1,050    |         1,050        |

jbschlosser commented on aten/src/ATen/native/cuda/ConvolutionMM2d.cu:
```diff
@@ -117,17 +117,13 @@ void slow_conv2d_forward(
   slow_conv2d_shape_check(
       input, {}, weight, bias, kH, kW, dH, dW, padH, padW, /*weight_nullable*/false);
 
-  TORCH_CHECK(!bias.defined() || bias.is_contiguous(),
-              "bias tensor has to be contiguous");
-
```
Just to verify my understading: `bias` no longer needs to be contiguous because `grad_bias` is now being computed using `at::sum_out` instead of with an ad-hoc kernel that required contiguous input?

peterbell10 replied:
That's the general idea. This is replacing cuBLAS library calls which don't support completely arbitrary strides, with `TensorIterator` kernels that do.

jbschlosser reviewed (commented):
Very nice work! Left a couple comments below

peterbell10 committed:
Update on "Remove fgrad_input from slow_conv2d"




[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9490,31 +9490,31 @@
     CUDA: slow_conv_transpose3d_backward_cuda
 
-- func: thnn_conv2d.out(Tensor self, Tensor weight, int[2] kernel_size, Tensor? bias=None, int[2] stride=1, int[2] padding=0, *, Tensor(a!) out) -> Tensor(a!)
+- func: _slow_conv2d.out(Tensor self, Tensor weight, int[2] kernel_size, Tensor? bias=None, int[2] stride=1, int[2] padding=0, *, Tensor(a!) out) -> Tensor(a!)
```
I've restored the name of the top-level function but I think changing the names of `_forward` and `_backward` is okay since the signature is changing either way.

peterbell10 committed:
Update on "Remove fgrad_input from slow_conv2d"




[ghstack-poisoned]

jbschlosser reviewed (approved):
LGTM! Thanks for the update :)

ngimel commented:
This looks awesome and I'm glad we don't need blas to add bias/compute bias grad, it's a leftover from caffe1 times.

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30830887).

jbschlosser commented:
Hey @peterbell10, do you mind doing a rebase for this? auto-rebase is failing internally

peterbell10 committed:
Update on "Remove fgrad_input from slow_conv2d"


Differential Revision: [D30830887](https://our.internmc.facebook.com/intern/diff/D30830887)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30830887).

jbschlosser commented:
FYI @peterbell10 there's still quite a bit of internal breakage happening due to the renaming. I'm looking into how much of it is unavoidable due to the signature change.

peterbell10 commented:
@jbschlosser any update on this?

jbschlosser commented:
Hey @peterbell10, sorry for the lack of updates. It's merged now.
